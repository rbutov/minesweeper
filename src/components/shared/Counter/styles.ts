import styled from 'styled-components';

import nums_background from 'assets/nums/nums_background.svg';

const CounterContainer = styled.div`
  width: 61.5px;
  height: 37.5px;
  background-size: 100% 100%;
  padding: 3px;
  background-image: url(${nums_background});

  > div {
    float: left;
    width: 33%;
  }
`;

export { CounterContainer };
