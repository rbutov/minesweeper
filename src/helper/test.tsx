import React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';
import { ReactWrapper, ShallowWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { confStore } from 'store';

type Wrapper<P = unknown, S = unknown> = ShallowWrapper<P, S> | ReactWrapper<P, S>;
type MountFnType<P, S> = (
  node: React.ReactElement<P>,
  options?: Record<string, unknown>
) => Wrapper<P, S>;

const wrapAppComponent = <P, S>(
  renderer: MountFnType<P, S>,
  element: JSX.Element,
  other?: {
    store?: Store;
  }
): Wrapper<P, S> => {
  const coreStore = confStore();

  return renderer(<Provider store={other?.store ? other.store : coreStore}>{element}</Provider>);
};

const waitForComponentToPaint = async <P, S>(
  wrapper: Wrapper<P, S>,
  other?: { delay?: number; props?: Pick<P, keyof P> }
) => {
  await act(async () => {
    await runAllPromises();
    await new Promise((resolve) => setTimeout(resolve, other?.delay ? other.delay : 0));
    wrapper.setProps({ ...other?.props });
    wrapper.update();
  });
};

const runAllPromises = async () => {
  return new Promise<void>((resolve) => {
    setTimeout(resolve, 0);
  });
};

export { wrapAppComponent, waitForComponentToPaint };
export type { Wrapper };
