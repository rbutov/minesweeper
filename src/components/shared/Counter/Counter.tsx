import React, { useEffect, useState } from 'react';
import { useAppSelector } from 'hooks/useAppSelector';
import { getMinesCount } from 'helper/game';
import { Number } from 'components/shared/Number';
import { CounterContainer } from './styles';

const Counter = () => {
  const [count, setCount] = useState([0, 0, 0]);
  const { minesCount, flaggedKeys } = useAppSelector((state) => state.field);

  useEffect(() => {
    setCount(getMinesCount({ minesCount, flaggedKeys }));
  }, [minesCount, flaggedKeys]);

  return (
    <CounterContainer>
      <Number num={count[0]} />
      <Number num={count[1]} />
      <Number num={count[2]} />
    </CounterContainer>
  );
};

export { Counter };
