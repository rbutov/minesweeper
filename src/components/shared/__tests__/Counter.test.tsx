import React from 'react';
import { wrapAppComponent } from 'helpers/test';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Counter } from '../Counter';

describe('<Counter />', () => {
  it('_render', () => {
    const wrapper = wrapAppComponent(mount, <Counter />);

    const numbers0 = wrapper.find({ 'data-testid': 'number-0' });
    expect(numbers0).toHaveLength(4);
    const numbers1 = wrapper.find({ 'data-testid': 'number-1' });
    expect(numbers1).toHaveLength(2);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
