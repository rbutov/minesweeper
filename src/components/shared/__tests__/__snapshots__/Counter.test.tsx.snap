// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> _render 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Counter>
    <styled.div
      data-testid="counter"
    >
      <div
        className="sc-hKwDye cHUlqo"
        data-testid="counter"
      >
        <Number
          num={0}
        >
          <styled.div
            data-testid="number-0"
            num={0}
          >
            <div
              className="sc-dkPtRN jWYRyZ"
              data-testid="number-0"
            />
          </styled.div>
        </Number>
        <Number
          num={1}
        >
          <styled.div
            data-testid="number-1"
            num={1}
          >
            <div
              className="sc-dkPtRN kQGVMA"
              data-testid="number-1"
            />
          </styled.div>
        </Number>
        <Number
          num={0}
        >
          <styled.div
            data-testid="number-0"
            num={0}
          >
            <div
              className="sc-dkPtRN jWYRyZ"
              data-testid="number-0"
            />
          </styled.div>
        </Number>
      </div>
    </styled.div>
  </Counter>
</Provider>
`;
